### PR TITLE
Biquad EQ with full kill feature

### PR DESCRIPTION
**Pull request by daschuer**
This is an attempt to build a new default EQ for Mixxx. It combines an standard Biquad Equalizer with an Isolator to reach full kill. This should suite to gentle tweak the sound around unity and It allows to full kill a band when doing a transition.

Features:
* Boost: +12 dB
* Kill: -Infinity 
* Bit perfect at unity  
* No CPU load at unity 
* Up medium CPU load when killing Mid 
* Gentle rounded cut-off

I am curious to read your results from listening.

### CODE REVIEW

File: src/effects/native/biquadfullkilleqeffect.h [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+    std::unique_ptr<EngineFilterBiquad1Peaking> m_midBoost;
+    std::unique_ptr<EngineFilterBiquad1Peaking> m_highBoost;
+    std::unique_ptr<EngineFilterBiquad1LowShelving> m_lowKill;
+    std::unique_ptr<EngineFilterBiquad1Peaking> m_midKill;
+    std::unique_ptr<EngineFilterBiquad1HighShelving> m_highKill;
+    std::unique_ptr<EngineFilterBessel4Low> m_low1;
+    std::unique_ptr<EngineFilterBessel4Low> m_low2;
+    std::unique_ptr<EngineFilterDelay<kMaxDelay2>> m_delay2;
+    std::unique_ptr<EngineFilterDelay<kMaxDelay2>> m_delay3;
+    CSAMPLE* m_pLowBuf;
```
Comment (uklotzde) [inline src/effects/native/biquadfullkilleqeffect.h]: Why don't you use SampleBuffer to avoid the manual alloc()/free()?

**Comment by Be-ing** [comment]
This colors the sound in a way that is similar to the Linkwitz-Riley ISO but not as prominent. I think there is some value to having the default EQ be full kill, so this would make a good default. Personally I prefer the transparency of the plain Biquad EQ. They all sure sound better than the Bessel ISOs.

**Comment by Be-ing** [comment]
Perhaps when we announce 2.1 beta there can be a poll on the blog asking what EQ users like best.

**Comment by daschuer** [comment]
> Personally I prefer the transparency of the plain Biquad EQ. 

Can you explain more with which settings this EQ is less  transparent? May goal was to achieve the same sound around unity and "just" expand the -23 dB kill to -Infinity dB

Maybe we can tweak some settings it to be more transparent, in critical settings. 

> Perhaps when we announce 2.1 beta there can be a poll on the blog asking what EQ users like best.

Good idea, we should also allow to wish additional types, to get a clue what is important. 

I am just afraid that the ISO users will vote for the LR8 and the EQ users will vote for the plain Biquad EQ. 
I assume that the majority of users has not a strong opinion to that. So my idea was to add this as a default, and cover both use-cases plus bit-perfect at unity if a user does not need EQs.

**Comment by Be-ing** [comment]
> Can you explain more with which settings this EQ is less transparent? May goal was to achieve the same sound around unity and "just" expand the -23 dB kill to -Infinity dB

I start to notice slight coloration if I listen closely around the 10 o'clock position. At and near center it sounds transparent.

I don't understand how the EQ and isolator are working together, so I'm not sure if this makes sense, but maybe you could change it so the isolator doesn't activate until the knob is closer to all the way down?

> Good idea, we should also allow to wish additional types, to get a clue what is important.

Yeah, including an "other" option is a good idea.

> I am just afraid that the ISO users will vote for the LR8 and the EQ users will vote for the plain Biquad EQ.

My main concern about the validity of the results would be if most users just vote for the default that we set for the beta and don't bother testing the various options. Btw, why is Loudness Contour marked as an EQ effect?

**Comment by daschuer** [comment]
The right knob side is the same as the Biquad EQ. 0 .. 12 dB 
The left side has three regions 
12:00 ...  10:00 pure Biquad EQ 0 ... -12 dB
10:00 ... 8:30 Biquad EQ -12 ... -23dB; Bessel4 0 .. -28 dB = -12 ... -51 dB 
8:30 ... 7:30 Biquad EQ -23dB; Bessel4 -28  .. -Inf dB = -51 ... -inf dB 

The pure Biquad EQ has a linear dB scale and has -12 dB @ 9:15 and -23 dB @Be-ing  7:30

**Comment by daschuer** [comment]
OK, I think I have Improved the situation. I have moved the ISO start region to 9:00 and -12 dB. Now we have a symetric +-12 dB pure biquad range, like our Graphic EQ. The rest of the scale is a combination of ISO and EQ. The EQ scale is scratched to void the change of resolution, like in the pervious version. 

Do you like it?

**Comment by Be-ing** [comment]
It is better. I wouldn't be opposed to it being the default, but I still prefer the Biquad and the Linkwitz-Riley. Would it be possible to combine those for a full kill effect? I presume the phase shift of the Linkwitz-Riley would present a challenge to work with another effect that does not produce a phase shift.

**Comment by daschuer** [comment]
Technical you could combine a Biquad EQ with a Linkwitz-Riley. As you can chain them together in a effect rack. But I am in doubt if that makes sense, because both types are just doing the opposite . The LR with it straight slopes and the BQ with its rounded curves. 

Reading your test results again, I  wonder why you was able to hear the 10:00 position in the first version and you are sensitive with the original Bessel around unity.  
From the theory, the Bessel is enabled there, but it should not alter the sound compared to the Bq which is already at -8 dB, so I assume, there must be an issue the ramping code.

How is your buffer size setting in setting hardware preferences? Which band is effected the most? 

Please test the following: 
* load the Bessel4 into a effect rack 
* Try to toggle the most effected band between 0.5 and 0.48 using one click on the mouse Wheel. Do you hear a difference? Do you hear the ramp? 
* move most effected Band to parameter = 0.48 (one Mouse wheel click)  
* Toggle the effect enable button. Do you hear a difference? Do you hear the ramp? 
* Now do the same test with all bands to parameter = 0.48. Results? 

Thank you.

**Comment by Be-ing** [comment]
I can't hear any difference between exact center and moving my mouse wheel once. But even with all knobs at center, enabling either the Bessel4 or Bessel8 produces an audible artifact. Same with the Linkwitz-Riley. Those are issues that should be fixed, but that's not what I was referring to in my previous comment. That little pop sound when enabling the effects is different from the coloration this effect produces.

**Comment by daschuer** [comment]
I am already working on the pop issue. 

Which band is effected the most by coloring? What is the clock position you are starting to hear it in the Bq EQ/ISO?

**Comment by daschuer** [comment]
The pop and click issues are gone now. 
I have made this one the default, until we know better. 

Ready for merge?

**Comment by Be-ing** [comment]
> Which band is effected the most by coloring?

I think that depends on the track.

> What is the clock position you are starting to hear it in the Bq EQ/ISO?

Around 9 o'clock.

> The pop and click issues are gone now.

Yes, they are for the Bessel ISOs, but the Linkwitz-Riley still pops when enabled.

>  I have made this one the default, until we know better.

Sure, we can solicit more feedback from the 2.1 betas.

**Comment by Be-ing** [comment]
I noticed this in my console output. I do not know what I was doing when it was printed:
```
Warning [Engine]: DEBUG ASSERT: "pInput != pOutput" in function void LVMixEQEffectGroupState<LPF>::processChannelAndPause(const CSAMPLE*, CSAMPLE*, int) [with LPF = EngineFilterBessel4Low; CSAMPLE = float] at src/effects/native/lvmixeqbase.h:163
```

**Comment by daschuer** [comment]
Ups, thats a leftover and must be removed. 
The clicking in the Linkwitz-Riley is a separate issue.


>> What is the clock position you are starting to hear it in the Bq EQ/ISO?

> Around 9 o'clock.

Interesting, that the point where the we switch from one EQ to ISO scale.
The ISO part Itself is still not notable there, proved in your earlier test, with the Bessel around unity.

**Comment by Be-ing** [comment]
To clarify, it's not very notable around 9 o'clock. That's where I start to hear it if I'm listening for it.

**Comment by daschuer** [comment]
Ah, OK, that means it normal, a coloring is expected since the filter slope moves a bit like a filter effect when coming into action.

**Comment by daschuer** [comment]
Thank you for review!

**Comment by ronso0** [comment]
I just checked out this new EQ mode, but I can't hear any effect. Biquad without Kill works fine.
Filed a bug [here](https://bugs.launchpad.net/mixxx/+bug/1667137)

**Pull request merged**